libraries/AP_RangeFinder/AP_RangeFinder_TOFSenseF_I2C.cpp

```diff
@@ -0,0 +1,156 @@
+#include <AP_HAL/utility/sparse-endian.h>
+
+extern const AP_HAL::HAL& hal;
+
+AP_RangeFinder_TOFSenseF_I2C::AP_RangeFinder_TOFSenseF_I2C(RangeFinder::RangeFinder_State &_state,
+                                                           AP_RangeFinder_Params &_params,
+                                                           AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_RangeFinder_Backend(_state, _params)
+    , _dev(std::move(dev))
+{
+}
+
+// detect if a TOFSenseP rangefinder is connected. We'll detect by
+// trying to take a reading on I2C. If we get a result the sensor is
+// there.
+AP_RangeFinder_Backend *AP_RangeFinder_TOFSenseF_I2C::detect(RangeFinder::RangeFinder_State &_state,
+																AP_RangeFinder_Params &_params,
+                                                             AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_RangeFinder_TOFSenseF_I2C *sensor
+        = new AP_RangeFinder_TOFSenseF_I2C(_state, _params, std::move(dev));
+    if (!sensor) {
+        return nullptr;
+    }
+
+    if (!sensor->_init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+// initialise sensor
+bool AP_RangeFinder_TOFSenseF_I2C::_init(void)
+{
+    _dev->get_semaphore()->take_blocking();
+
+    if (!start_reading()) {
+        _dev->get_semaphore()->give();
+        return false;
+    }
+
+    // give time for the sensor to process the request
+    hal.scheduler->delay(100);
+
+    uint32_t reading_mm;
+    uint16_t status;
+    uint16_t signal_strength;
+
+    if (!get_reading(reading_mm, signal_strength, status)) {
+        _dev->get_semaphore()->give();
+        return false;
+    }
+
+    _dev->get_semaphore()->give();
+
+    _dev->register_periodic_callback(100000,
+                                     FUNCTOR_BIND_MEMBER(&AP_RangeFinder_TOFSenseF_I2C::_timer, void));
+
+    return true;
+}
+
+// start_reading() - ask sensor to make a range reading
+bool AP_RangeFinder_TOFSenseF_I2C::start_reading()
+{
+    uint8_t cmd[] = {TOFSENSEP_I2C_COMMAND_TAKE_RANGE_READING, TOFSENSEP_I2C_COMMAND_SIGNAL_STATUS};
+
+    // send command to take reading
+    return _dev->transfer(cmd, sizeof(cmd), nullptr, 0);
+}
+
+// read - return last value measured by sensor
+bool AP_RangeFinder_TOFSenseF_I2C::get_reading(uint32_t &reading_mm, uint16_t &signal_strength, uint16_t &status)
+{
+
+    struct PACKED {
+        le32_t distance_mm;
+        uint32_t signal_strength_and_status;
+    } packet;
+
+    // take range reading and read back results
+    const bool ret = _dev->transfer(nullptr, 0, (uint8_t *) &packet, sizeof(packet));
+
+    if (ret) {
+        // combine results into distance
+        reading_mm = le32toh(packet.distance_mm);
+        signal_strength = (uint16_t)(packet.signal_strength_and_status >> 16);
+        status = (uint16_t)(packet.signal_strength_and_status);
+
+        // trigger a new reading
+        start_reading();
```

Comment:
this looks like if transfer fails, then no new start_reading() is done ever again

Reply:
same bug in maxsonar i2c